File: src/audioWfc/musicTheory/chords/Chord.java
package audioWfc.musicTheory.chords;

import audioWfc.musicTheory.Note;
import audioWfc.musicTheory.NoteSet;

import java.text.ParseException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;

import static audioWfc.musicTheory.chords.ChordQuality.MAJOR;
import static audioWfc.musicTheory.chords.ChordQuality.MINOR;

public abstract class Chord extends NoteSet {
    protected Note third;
    protected Note fifth;

    abstract String notation();

    abstract ChordQuality quality();

    @Override
    public String toString() {
        return "Chord: " + root + notation();
    }

    public Note getThird() {
        return third;
    }

    public Note getFifth() {
        return fifth;
    }

    public static Chord create(Note root, ChordQuality quality){
        switch (quality){
            case MAJOR -> {return new MajorChord(root);}
            case MINOR -> {return new MinorChord(root);}
            case DIMINISHED -> {return new DiminishedChord(root);}
            case AUGMENTED -> {return new AugmentedChord(root);}
            default -> {throw new RuntimeException("Unknown chord quality");}
        }
    }

    public static Set<Chord> getAllBasicChords(){
        Set<Chord> out = new HashSet<>();
        for (Note root : Note.values()){
            for (ChordQuality quality : Set.of(MAJOR, MINOR)){
                out.add(create(root, quality));
            }
        }
        return out;
    }

    public static Chord parse(String s) {
        Note rootNote;
        ChordQuality chordQuality;

        s = s.trim();

        // Parse the root note from the chord string
        if (s.length() == 1) {
            return new MajorChord(Note.valueOf(s.toUpperCase(Locale.ROOT)));
        } else if (s.length() == 2 && s.charAt(1) == 'm'){
            return new MinorChord(Note.valueOf(String.valueOf(Character.toUpperCase(s.charAt(0)))));
        } else if (s.charAt(1) == '#') {
            String rootNoteStr = s.substring(0, 2).replace("#", "S").toUpperCase(Locale.ROOT);
            rootNote = Note.valueOf(rootNoteStr);
            if (s.length() == 2) {
                return new MajorChord(rootNote);
            } else if (s.length() == 3 && (s.charAt(2) == 'm')) {
                return new MinorChord(rootNote);
            }
        }
        throw new RuntimeException("Invalid chord:" + s);
    }

    public static Set<Chord> parseSet(String s) {
        String[] chordStrings = s.trim().split("\\s+");
        return Arrays.stream(chordStrings)
                .map(Chord::parse)
                .collect(Collectors.toSet());
    }



}

File: src/music_theory/Chord.ts
import { arrayEquals } from "../util/utils"
import { Chordesque, ChordesqueIR } from "../wfc/hierarchy/Chordesque"
import { Section } from "../wfc/hierarchy/Section"
import { Note, relativeNote } from "./Note"
import { NoteSet } from "./NoteSet"

export type ConcreteChordQuality =
	| "major"
	| "minor"
	| "diminished"
	| "augmented"
export type ChordQuality = ConcreteChordQuality | null
export type ChordIR = {
	root: Note
	quality: ChordQuality
}

export const isChordIR = (obj: ChordesqueIR): obj is ChordIR => {
	return typeof obj == "object"
}

export function stringToChordIR(chordString: string) {
	if (!["A", "B", "C", "D", "E", "F", "G"].includes(chordString[0]))
		return undefined

	const endOfRoot = chordString[1] == "#" ? 2 : 1
	const root = chordString.slice(0, endOfRoot) as Note
	const quality = chordString.slice(endOfRoot)

	if (["", "m", "°", "+"].includes(quality))
		return { root: root, quality: notationToQuality(quality)! }

	return undefined
}

export function chordIRToString(chordIR: ChordIR) {
	return `${chordIR.root}${qualityToNotation(chordIR.quality)}`
}

function notationToQuality(notation: string): ChordQuality | undefined {
	switch (notation) {
		case "":
			return "major"
		case "m":
			return "minor"
		case "°":
			return "diminished"
		case "+":
			return "augmented"
		default:
			return undefined
	}
}

function qualityToNotation(quality: ChordQuality): string {
	switch (quality) {
		case "major":
			return ""
		case "minor":
			return "m"
		case "diminished":
			return "°"
		case "augmented":
			return "+"
		default:
			throw new Error("Invalid chord quality")
	}
}

export class Chord extends NoteSet implements Chordesque {
	protected third: Note
	protected fifth: Note
	protected extensions?: Note[]

	constructor(root: Note, noteValues: number[]) {
		super(root, noteValues)
		this.third = relativeNote(root, noteValues[1])
		this.fifth = relativeNote(root, noteValues[2])
		this.extensions = noteValues
			.slice(3)
			.map((value) => relativeNote(root, value))
	}

	static fromRootAndQuality(root: Note, quality: ChordQuality): Chord {
		switch (quality) {
			case "major":
				return new MajorChord(root)
			case "minor":
				return new MinorChord(root)
			case "diminished":
				return new DiminishedChord(root)
			case "augmented":
				return new AugmentedChord(root)
			default:
				throw new Error("Invalid chord quality: " + quality)
		}
	}

	static fromIR(chordIR: ChordIR): Chord {
		return Chord.fromRootAndQuality(chordIR.root, chordIR.quality)
	}

	getChord() {
		return this
	}

	getBpm(section: Section){
		return section.getObj().bpm
	}

	static allBasicChords(): Chord[] {
		const chords: Chord[] = []
		for (const note of Object.values(Note)) {
			chords.push(new MajorChord(note))
			chords.push(new MinorChord(note))
		}
		return chords
	}

	equals(other: Chord): boolean {
		return (
			this.root == other.root &&
			arrayEquals(this.noteValues, other.noteValues)
		)
	}

	getThird() {
		return this.third
	}

	getFifth() {
		return this.fifth
	}

	toString() {
		return `${this.root} ${this.noteValues
			.map((value) => `+${value}`)
			.join("")}`
	}

	getName() {
		return this.toString()
	}

	static parseChordString(chordString: string): Chord | undefined {
		const chordIR = stringToChordIR(chordString)
		if (chordIR == undefined) return undefined
		return Chord.fromRootAndQuality(chordIR.root, chordIR.quality)
	}

	static parseChordsString(chordsString: string): Chord[] | undefined {
		if (chordsString == "") return []
		const res = chordsString
			.trim()
			.split(" ")
			.map((chordString) => Chord.parseChordString(chordString))
		if (res.includes(undefined)) return undefined
		return res as Chord[]
	}

	clone(): Chordesque {
		return new Chord(this.root, this.noteValues)
	}
}

export class MajorChord extends Chord {
	constructor(root: Note) {
		super(root, [0, 4, 7])
	}

	toString() {
		return `${this.root}`
	}
}

export class MinorChord extends Chord {
	constructor(root: Note) {
		super(root, [0, 3, 7])
	}

	toString() {
		return `${this.root}m`
	}
}

export class DiminishedChord extends Chord {
	constructor(root: Note) {
		super(root, [0, 3, 6])
	}

	toString() {
		return `${this.root}°`
	}
}

export class AugmentedChord extends Chord {
	constructor(root: Note) {
		super(root, [0, 4, 8])
	}

	toString() {
		return `${this.root}+`
	}
}
